attester/src/sgx_dcap/mod.rs

```diff
@@ -0,0 +1,102 @@
+// Copyright (c) 2023 Alibaba Cloud
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use super::Attester;
+use anyhow::{anyhow, bail, Result};
+use occlum_dcap::{sgx_report_data_t, DcapQuote};
+use serde::{Deserialize, Serialize};
+
+const OCCLUM_ENV: &str = "OCCLUM";
+
+enum SgxLibOsType {
+    Invalid,
+    Occlum,
+    Gramine,
+}
+
+fn get_libos_type() -> SgxLibOsType {
+    if std::env::var(OCCLUM_ENV).is_ok() {
+        return SgxLibOsType::Occlum;
+    }
+
+    match std::fs::read_to_string("/dev/attestation/attestation_type") {
+        Ok(d) if d == "dcap" => SgxLibOsType::Gramine,
+        _ => SgxLibOsType::Invalid,
+    }
+}
+
+pub fn detect_platform() -> bool {
+    match get_libos_type() {
+        SgxLibOsType::Invalid => false,
+        SgxLibOsType::Occlum => true,
+        SgxLibOsType::Gramine => true,
+    }
+}
+
+#[derive(Serialize, Deserialize)]
+struct SgxDcapAttesterEvidence {
+    /// Base64 encoded SGX quote.
+    quote: String,
```

Comment:
Do we need to add another field here named `libos_type` to indicate the type of libos? Though the integrity of this field is not protected.

Reply:
In my first version I had it for `SgxDcapAttester` and then `Defaulter` implemented for it that initialized it but then dropped that version. The verifier can use `quote` regardless of who generated it.

Reply:
Well. I'm thinking about how the verifier choose the golden value of the MRENCLAVE.
1. With extra `libos_type`: There should be two lists of golden values for occlum and gramine agent-enclave, and they can be queried due to different `libos_type`
2. We only keep one list of both golden values for occlum and gramine agent-enclave. When verifying the quote, attestation service will use the values in the list to try to match the one in quote.

Both are fine with me.

Reply:
OK I see what you mean. I'd keep this PR as it is now and think about the quote verification for Gramine/Occlum co-existence later.